src/zepben/evolve/services/network/tracing/networktrace/actions/equipment_tree_builder.py

```diff
@@ -66,6 +67,7 @@ def _apply(self, item: NetworkTraceStep[Any], context: StepContext):
         current_node: TreeNode = self.get_context_value(context)
         if current_node.parent:
             current_node.parent.add_child(current_node)
+        self.leaves[current_node.identified_object.mrid] = current_node
```

Comment:
doesn't this just specify every node as a leaf? Leaves should only be the ends of branches (i.e nodes with only one branch). 

This is probably not desirable either as it ends up using a lot more memory for something that is not hard to compute, and will only be used rarely. There is a few corrections/optimisations to make here:

1. Leaves can be classified as any node that has only upstream terminals.
2. This can be computed on the result, so make self.leaves a lazy calculation - i.e turn it into a @property that just looks for ConductingEquipment with only upstream terminals in the entire tree and then saves that to self.leaves and returns it.

This way we would only calculate and store the leaves if the user actually asked for it.

Reply:
addressed, theres an edge case with loops.